pkg/antctl/raw/bundle/command.go

```diff
@@ -0,0 +1,507 @@
+
+	agentapiserver "github.com/vmware-tanzu/antrea/pkg/agent/apiserver"
+	"github.com/vmware-tanzu/antrea/pkg/agent/controller/noderoute"
+	"github.com/vmware-tanzu/antrea/pkg/antctl/runtime"
+	systemv1beta1 "github.com/vmware-tanzu/antrea/pkg/apis/system/v1beta1"
+	controllerapiserver "github.com/vmware-tanzu/antrea/pkg/apiserver"
+	antrea "github.com/vmware-tanzu/antrea/pkg/client/clientset/versioned"
+)
+
+const (
+	barTmpl      pb.ProgressBarTemplate = `{{string . "prefix"}}{{bar . }} {{percent . }} {{rtime . "ETA %s"}}` // Example: 'Prefix[-->______] 20%'
+	requestRate                         = 100
+	requestBurst                        = 150
+)
+
+// Command is the bundle command implementation.
+var Command *cobra.Command
+
+var option = &struct {
+	dir           string
+	labelSelector string
+}{}
+
+var remoteControllerLongDescription = strings.TrimSpace(`
+	Generate support bundles of the cluster, includes: antrea agent bundles, antrea controller bundles and general information of the cluster.
+`)
+
+var remoteControllerExample = strings.Trim(`
+  Generate support bundle of all nodes and save them to current working dir
+  $ antctl bundle
+  Generate support bundle of specific nodes that filtered by wildcard
+  $ antctl bundle *worker*
+  Generate support bundle of specific nodes that filtered by wildcard and label selectors
+  $ antctl bundle *worker* -l kubernetes.io/os=linux
+  Generate support bundle of all nodes and save them to specific dir
+  $ antctl bundle -d ~/Downloads
+`, "\n")
+
+func init() {
+	Command = &cobra.Command{
+		Use:   "bundle",
+		Short: "Generate support bundle",
+	}
+
+	if runtime.Mode == runtime.ModeAgent {
+		Command.RunE = agentRunE
+		Command.Long = "Generate the support bundle of current antrea agent."
+	} else if runtime.Mode == runtime.ModeController && runtime.InPod {
+		Command.RunE = controllerLocalRunE
+		Command.Long = "Generate the support bundle of current antrea controller."
+	} else if runtime.Mode == runtime.ModeController && !runtime.InPod {
+		Command.Args = cobra.MaximumNArgs(1)
+		Command.Use += " [nodeName]"
+		Command.Long = remoteControllerLongDescription
+		Command.Example = remoteControllerExample
+		cwd, _ := os.Getwd()
+		Command.Flags().StringVarP(&option.dir, "dir", "d", filepath.Join(cwd, "bundles"), "bundle output dir, the path will be created if it not exists")
+		Command.Flags().StringVarP(&option.labelSelector, "label-selector", "l", "", "selector (label query) to filter on nodes, supports '=', '==', and '!='.(e.g. -l key1=value1,key2=value2)")
+		Command.RunE = controllerRemoteRunE
+	}
+}
+
+func setupKubeconfig(kubeconfig *rest.Config) {
+	kubeconfig.APIPath = "/apis"
+	kubeconfig.GroupVersion = &systemv1beta1.SchemeGroupVersion
+	kubeconfig.NegotiatedSerializer = scheme.Codecs.WithoutConversion()
+	// TODO: enable secure connection.
+	kubeconfig.Insecure = true
+	kubeconfig.CAFile = ""
+	kubeconfig.CAData = nil
+	if runtime.InPod {
+		port, ok := os.LookupEnv("API_PORT")
```

Comment:
I thought you added it to CRD? User will need to update several places if they change the port.

Reply:
The CRD is for out-of-cluster usage in my purpose.
Although we could still try to connect to the default port and let users specify the correct one if needed, it is convenient for users when the tool can choose automatically.

Reply:
But then it's not convenient for user to specify a different APIPort when deploying as there are 3 places need to be updated? And it's not documented.
We don't have a choice for the current two places, I would suggest not introducing one more redundant config and suggest optimizing the port auto-discovery if no good approach for now.

Reply:
OK, I reverted the environment variable change for now.
Also, I added a TODO comment for generating a kubeconfig which contains token and port information in Pod for antctl.